File: java/organizing-policy/src/main/java/com/hexagonalchess/organizing/StartGame.java
package com.hexagonalchess.organizing;

public class StartGame {

    private final GameRepository gameRepository;
    private final GameInitializer gameInitializer;

    public StartGame(GameRepository gameRepository, GameInitializer gameInitializer) {
        this.gameRepository = gameRepository;
        this.gameInitializer = gameInitializer;
    }

    public GameId execute(String whiteName, String blackName) {
        var white = new Player(whiteName);
        var black = new Player(blackName);
        var gameId = gameInitializer.initializeGame(white, black);
        gameRepository.save(new Game(gameId, white, black));
        return gameId;
    }
}

File: ruby/organizing_policy/lib/organizing/start_game.rb
require_relative 'game_id'
require_relative 'player'
require_relative 'game'
require_relative 'game_repository'
require_relative 'game_initializer'

module Organizing
  class StartGame
    def initialize(game_repository, game_initializer)
      @game_repository  = game_repository
      @game_initializer = game_initializer
    end

    def call(white_name, black_name)
      white = Player.new(name: white_name)
      black = Player.new(name: black_name)
      game_id = @game_initializer.initialize_game(white, black)
      @game_repository.save(Game.new(id: game_id, white: white, black: black))
      game_id
    end
  end
end
